File: benchmarks/pandas/bench_dataframe_ffill_bfill_fn.py
"""
Benchmark: pandas DataFrame.ffill() / DataFrame.bfill() — forward/backward fill.
Mirrors tsb bench_dataframe_ffill_bfill_fn.ts (standalone dataFrameFfill/dataFrameBfill).
Outputs JSON: {"function": "dataframe_ffill_bfill_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

rng = np.random.default_rng(42)
a = np.where(np.arange(SIZE) % 5 == 0, np.nan, np.arange(SIZE, dtype=float))
b = np.where(np.arange(SIZE) % 7 == 0, np.nan, np.arange(SIZE, dtype=float) * 0.5)
c = np.where(np.arange(SIZE) % 3 == 0, np.nan, np.arange(SIZE, dtype=float) * 2.0)

df = pd.DataFrame({"a": a, "b": b, "c": c})

for _ in range(WARMUP):
    df.ffill()
    df.bfill()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.ffill()
    df.bfill()
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "dataframe_ffill_bfill_fn",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))

File: benchmarks/tsb/bench_dataframe_ffill_bfill_fn.ts
/**
 * Benchmark: dataFrameFfill / dataFrameBfill — standalone forward/backward fill for DataFrames.
 * Mirrors pandas DataFrame.ffill() / DataFrame.bfill().
 * Outputs JSON: {"function": "dataframe_ffill_bfill_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, dataFrameFfill, dataFrameBfill } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

const df = DataFrame.fromColumns({
  a: Array.from({ length: SIZE }, (_, i) => (i % 5 === 0 ? null : i * 1.0)),
  b: Array.from({ length: SIZE }, (_, i) => (i % 7 === 0 ? null : i * 0.5)),
  c: Array.from({ length: SIZE }, (_, i) => (i % 3 === 0 ? null : i * 2.0)),
});

for (let i = 0; i < WARMUP; i++) {
  dataFrameFfill(df);
  dataFrameBfill(df);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  dataFrameFfill(df);
  dataFrameBfill(df);
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "dataframe_ffill_bfill_fn",
    mean_ms: round3(total / ITERATIONS),
    iterations: ITERATIONS,
    total_ms: round3(total),
  }),
);

function round3(v: number): number {
  return Math.round(v * 1000) / 1000;
}
